### PR TITLE
8276123: ZipFile::getEntry will not return a file entry when there is a directory entry of the same name within a Zip File

### PR DESCRIPTION
**Pull request by LanceAndersen**
Hi all,

This patch addresses a regression introduced in JDK 15 via JDK-8242959 where you can no longer access a file entry contained within a Zip file when there is also a directory entry with the same name via ZipFile:getEntry(). 

 Once fixed, the behavior will be consistent with earlier JDK releases. 

Mach5 tiers 1-3 have been run without failure

Best
Lance

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issue
 * [JDK-8276123](https://bugs.openjdk.java.net/browse/JDK-8276123): ZipFile::getEntry will not return a file entry when there is a directory entry of the same name within a Zip File

### Reviewers
 * [Claes Redestad](https://openjdk.java.net/census#redestad) (@cl4es - **Reviewer**)
 * [Alan Bateman](https://openjdk.java.net/census#alanb) (@AlanBateman - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.java.net/jdk pull/6342/head:pull/6342` \
`$ git checkout pull/6342`

Update a local copy of the PR: \
`$ git checkout pull/6342` \
`$ git pull https://git.openjdk.java.net/jdk pull/6342/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 6342`

View PR using the GUI difftool: \
`$ git pr show -t 6342`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.java.net/jdk/pull/6342.diff">https://git.openjdk.java.net/jdk/pull/6342.diff</a>

</details>

### CODE REVIEW

File: test/jdk/java/util/zip/ZipFile/ZipFileDuplicateEntryTest.java [hunk 1]
```diff
@@ -0,0 +1,585 @@
...
+        ByteArrayOutputStream buffer = new ByteArrayOutputStream();
+        int len;
+        while ((len = is.read(data, 0, data.length)) != -1) {
+            buffer.write(data, 0, len);
+        }
+        buffer.flush();
+        return buffer.toByteArray();
+    }
+
+    /**
```
Comment (cl4es) [inline test/jdk/java/util/zip/ZipFile/ZipFileDuplicateEntryTest.java:515]: Left-over?

**Review by cl4es — APPROVED** [review]
LGTM!

File: test/jdk/java/util/zip/ZipFile/ZipFileDuplicateEntryTest.java (hunk 1 above)
Comment (LanceAndersen) [reply test/jdk/java/util/zip/ZipFile/ZipFileDuplicateEntryTest.java:515]: Thank you for the quick review Claes!

I was going to remove this method but chose to leave it in to make it a bit more straight forward to port to JDK 8 (which I have run the test on as well).  No strong preferences, just figured leaving it there might make it slightly easier for a back port.

File: src/java.base/share/classes/java/util/zip/ZipFile.java [hunk 2]
```diff
@@ -1631,13 +1631,18 @@ private int getEntryPos(String name, boolean addSlash) {
...
-                                entry.charAt(entryLen - 1) == '/')) {
+                        if (entryLen == nameLen && entry.equals(name)) {
+                            // Found our match
                             return pos;
                         }
+                        // If addSlash is true we'll now test for name+/ providing
+                        if (addSlash && nameLen + 1 == entryLen
+                                && entry.startsWith(name) &&
+                                entry.charAt(entryLen - 1) == '/') {
+                            // Now check for a match with a trailing slash
```
Comment (jaikiran) [inline src/java.base/share/classes/java/util/zip/ZipFile.java:1642]: Hello Lance,
Is this a typo in that comment? Should it instead say "... without a trailing slash"?

Comment (LanceAndersen) [reply src/java.base/share/classes/java/util/zip/ZipFile.java:1642]: Hi Jaikiran,

The comment is correct

File: test/jdk/java/util/zip/ZipFile/ZipFileDuplicateEntryTest.java [hunk 3]
```diff
@@ -0,0 +1,585 @@
...
+        Files.deleteIfExists(ZIP_FILE);
+        try (ZipOutputStream zos = new ZipOutputStream(Files.newOutputStream(ZIP_FILE))) {
+            zos.putNextEntry(new ZipEntry(FILE_ENTRY.name));
+            zos.write(FILE_ENTRY.bytes);
+            zos.closeEntry();
+            zos.putNextEntry(new ZipEntry(DIR_ENTRY.name));
+            zos.write(DIR_ENTRY.bytes);
+            zos.closeEntry();
+        } catch (IOException e) {
+            e.printStackTrace();
```
Comment (AlanBateman) [inline test/jdk/java/util/zip/ZipFile/ZipFileDuplicateEntryTest.java:176]: Did you mean to catch IOException here? If the setup fails then the test will still run.

File: src/java.base/share/classes/java/util/zip/ZipFile.java (hunk 2 above)
Comment (jaikiran) [reply src/java.base/share/classes/java/util/zip/ZipFile.java:1642]: Thank you for that clarification.  The "addSlash" param being "false" in the call below that comment is what made me think that the comment had a typo. I read that code in a bit more detail now and I see what that comment means. It essentially says that it's now issuing a call to find a entry position (if it exists) for the name ending with a "/" character.

Comment (cl4es) [reply src/java.base/share/classes/java/util/zip/ZipFile.java:1642]: The comment is a bit too fuzzy. What's happening here is we searched for 'foo' and found a match for `foo/` - but we can't be sure there's no `foo` entry later on in the hash bucket, so we (somewhat inefficiently) redo the search without the trailing slash.

I suggest clarifying to something like this:
```  // "name+/" entry found, keep looking for and prefer an exact match```

Comment (LanceAndersen) [reply src/java.base/share/classes/java/util/zip/ZipFile.java]: I just tweaked the comment slightly

File: test/jdk/java/util/zip/ZipFile/ZipFileDuplicateEntryTest.java (hunk 3 above)
Comment (LanceAndersen) [reply test/jdk/java/util/zip/ZipFile/ZipFileDuplicateEntryTest.java]: Removed the catch clause

**Review by cl4es — APPROVED** [review]

**Review by AlanBateman — APPROVED** [review]

**Pull request closed without merging**